ocaml/xapi/test_vdi_cbt.ml

```
@@ -209,6 +223,58 @@ let test_allowed_operations_updated_when_necessary () =
+    let _: _ API.Ref.t = Test_common.make_pbd ~__context ~host ~sR ~currently_attached:true () in
+    let vDI = Test_common.make_vdi ~__context ~is_a_snapshot:true ~managed:true ~cbt_enabled:true ~_type:`suspend ~sR () in
+    let vM = Test_common.make_vm ~__context () in
+    let _: _ API.Ref.t = Test_common.make_vbd ~__context ~uuid:"VBD-1" ~vDI ~vM ~currently_attached:false () in
+    register_smapiv2_server ~vdi_data_destroy:(fun _ ~dbg ~sr ~vdi -> ()) (Db.SR.get_uuid ~__context ~self:sR);
+
+    let check_vdi_is_snapshot_and_type ~vDI ~snapshot ~vdi_type =
+      OUnit.assert_equal ~msg:("VDI type should be set to " ^ (Record_util.vdi_type_to_string vdi_type))
+        (Db.VDI.get_type ~__context ~self:vDI) vdi_type;
+      let word = if snapshot then "" else " not" in
+      OUnit.assert_equal ~msg:("VDI should" ^ word ^ " be a snapshot")
+        (Db.VDI.get_is_a_snapshot ~__context ~self:vDI) snapshot
+    in
+    check_vdi_is_snapshot_and_type ~vDI ~snapshot:true ~vdi_type:`suspend;
+
+    (* set vDI as the suspend VDI of vM *)
+    Db.VM.set_suspend_VDI ~__context ~self:vM ~value:vDI;
+    OUnit.assert_equal ~msg:"VM.suspend_VDI should point to previously created VDI"
+      (Db.VM.get_suspend_VDI ~__context ~self:vM) vDI;
+
+    OUnit.assert_equal ~msg:"VDI should link to previously created VBD"
+      (List.map (fun vbd -> Db.VBD.get_uuid ~__context ~self:vbd)
+         (Db.VDI.get_VBDs ~__context ~self:vDI)) ["VBD-1"];
+
+    (* run VDI.data_destroy, check it has updated VDI fields *)
+    Xapi_vdi.data_destroy ~__context ~self:vDI;
+
+    OUnit.assert_equal ~msg:"VDI.data_destroy should set VDI type to cbt_metadata"
+      (Db.VDI.get_type ~__context ~self:vDI) `cbt_metadata;
+
+    OUnit.assert_equal ~msg:"VDI.data_destroy should destroy all associated VBDs"
+      (Db.VDI.get_VBDs ~__context ~self:vDI) [];
+
+    OUnit.assert_equal ~msg:"VM.suspend_VDI should be set to null"
+      (Db.VM.get_suspend_VDI ~__context ~self:vM) Ref.null;
```

One more test that we could add is to check that it throws `VDI_NOT_MANAGED` for a VDI that is not `managed` - last time we could not add it to the previous PR, because the check is done in the implementation.

---

I've now included this in the refactored `make_mock_server_infrastructure` function, it should throw the required error.

---

Actually, what we want to check here is not that the managed field is preserved, but if we call `data_destroy` on a VDI with `managed = false`, then it will throw `VDI_NOT_MANAGED`.